### PR TITLE
feat: Add ringbuffer support to libbpfgo

### PR DESCRIPTION
**Pull request by grantseltzer**
This adds support for ringbuffers to libbpfgo. A good reference for ringbuffers can be found [here](https://nakryiko.com/posts/bpf-ringbuf/). There is not much difference in userspace between ringbuffers and perfbuffers besides the fact that ringbuffers API does not support a callback for lost events. That will have to be handled in kernel space for consumers of libbpfgo, like Tracee. 

I have a separate PR awaiting this which I can open as a draft seeing how it would be consumed in Tracee.

### CODE REVIEW

File: libbpfgo/libbpfgo.go
Comment (grantseltzer) [inline libbpfgo/libbpfgo.go:42]: fixing whitespace right now

Comment (simar7) [reply libbpfgo/libbpfgo.go:42]: This brings up a good point though - how should we format the C code? We talked about Go code formatting but I'm curious to hear what you all are using.

File: libbpfgo/libbpfgo.go
Comment (yanivagman) [inline libbpfgo/libbpfgo.go:283]: extra newline

File: libbpfgo/libbpfgo.go
Comment (grantseltzer) [reply libbpfgo/libbpfgo.go:39]: Just to be consistent with the perf buffer init function, and to avoid passing `C.ringBufferCallback`. I can get rid of it if you feel strongly about it.

Comment (yanivagman) [reply libbpfgo/libbpfgo.go:39]: I think it is better to avoid unnecessary wrappers and use go where possible to make code more readable.
Not a must though

File: libbpfgo/libbpf_cb.go
Comment (grantseltzer) [reply libbpfgo/libbpf_cb.go]: Fixed back to int

**Comment by grantseltzer** [comment]
Marking this as ready for review, following up with another PR for selftests that include a runner script for it that I would subsequently add to github actions workflow.

**Comment by grantseltzer** [comment]
I actually just found a deadlock in this and the perfbuffer code when `Stop()` is called and the thread is blocked waiting for the event polling. Working on fix.

**Comment by grantseltzer** [comment]
Ok fixed

**Comment by yanivagman** [comment]
About you perfbuffer deadlock comment -
How can the perf poll loop be blocked if we have a timeout there?

File: libbpfgo/libbpfgo.go
Comment (yanivagman) [inline libbpfgo/libbpfgo.go:677]: Why not using timeout instead? (like we do for the perf buffer)
`err := C.ring_buffer__poll(rb.rb, 300)`

Comment (grantseltzer) [reply libbpfgo/libbpfgo.go:677]: I'm trying to figure out what is the exact unexpected behavior that's causing `ring_buffer__poll` to block regardless of the specified timeout. The test program I wrote exposed a block where the write to `rb.stop` in ringbuffer's `Stop` method to block indefinitely. Perhaps a channel was closed somewhere that the poll function was caught trying to write to. Perhaps a reused map file descriptor that's not being properly cleaned up. 

Pushing up that test program now so you can take a look.

**Comment by grantseltzer** [comment]
Alright so I just pushed a commit that I'd like you to look at @yanivagman 

The issue with having `C.ring_buffer__poll(rb.rb, 300)` (with the 300ms timeout) is that the ringbuffer can be free'd while waiting for an event to come in which causes a panic. The issue with my first approach of having the poll run in its own go routine is the same one, poll after free. 

I've removed the timeout from `ring_buffer__poll` (passing 0 for timeout) which seems to effectively solve the issue of getting caught with a free while polling.

**Comment by yanivagman** [comment]
Ok. I think I understand what you say.
So the perfbuffer can be freed by the Close() function while it is still polling - because the 300ms timeout didn't expire.
The problem here is that the pb.Stop() function didn't have the chance to gracefully stop the polling, right?
Changing the timeout to 0 will cause a busy loop, which will consume CPU resources, so it is not the best way to resolve that.
Instead, what we can do, is to introduce a new boolean variable to the PerfBuffer struct called "stopped".
We can then set this variable to true in the` case <-pb.stop:`. Then, we can wait with the Close() function (which frees the perf buffer) until we see that the stopped variable was set to true.
WDYT?

**Comment by grantseltzer** [comment]
@yanivagman I thought of this, but the problem there is there's still a chance for poll after free when the free happens after the check but before the poll call. Only way to safely do it that way is to use a lock which we want to avoid.

**Comment by yanivagman** [comment]
@grantseltzer I don't understand why there is still a chance for a poll after free if we implement this suggestion.
If you set the `stopped` variable in the` case <-pb.stop:`, no poll can happen after this point as you return from the poll loop right after setting the `stopped` variable

**Comment by grantseltzer** [comment]
Ah that's clever, I misunderstood your previous comment. I'll implement exactly this.

**Comment by grantseltzer** [comment]
@yanivagman found a bit of an issue with this approach as well. The reason `perf_buffer__poll` can hang is because channel sends block if there aren't any receivers. In the perf/ring buffer callbacks the the send will sit forever if the consumer of the events channel has stopped receiving on it. 

For example:

In the selftest:
```
recvLoop:
	for {
		 _ = <-eventsChannel     // events sent to eventsChannel in the callback
        	numberOfEventsReceived++ 
	        if numberOfEventsReceived > 3 {
	 	    break recvLoop	
	        }
	}
```

In this case when events are received after the loop is broken the callback will hang indefinitely (the timeout in `perf_buffer__poll` will not interrupt it). Therefore the stop signal is never received in your select statement. 

The fix here is to use a select statement in the callback which will make the write non-blocking. The other thing i've added is a check to see if `Close()` has been called multiple times to avoid a segfault.

File: libbpfgo/selftest/ringbuffers/.gitignore
Comment (itaysk) [inline libbpfgo/selftest/ringbuffers/.gitignore:2]: depending on the response to my other comments about libbpf.a and vmlinux.h , maybe need to add them here as well

File: libbpfgo/libbpfgo.go
Comment (itaysk) [inline libbpfgo/libbpfgo.go:660]: good catch, thinking about that what happens if Close is called before Stop. In this case you chose to wait until the buffer is closed, but I think an easier and reasonable approach, is to just call Stop? either way I don't think close should block the caller

Comment (itaysk) [reply libbpfgo/libbpfgo.go:660]: also, whatever we decide here probably applies to perf buffer (but can be implemented in a separate PR) cc: @yanivagman

File: libbpfgo/libbpfgo.go
Comment (itaysk) [reply libbpfgo/libbpfgo.go]: didn't really investigate this but just noting that it's impossible to write to a closed channel, more likely it's a double write to an unbuffered channel

File: libbpfgo/libbpfgo.go
Comment (itaysk) [reply libbpfgo/libbpfgo.go:660]: ok after posting the review I see that this was already discussed so I'll take a moment to re-read the history first

File: libbpfgo/selftest/ringbuffers/go.mod
Comment (grantseltzer) [reply libbpfgo/selftest/ringbuffers/go.mod:5]: Yes :-)

File: libbpfgo/selftest/ringbuffers/Makefile
Comment (grantseltzer) [reply libbpfgo/selftest/ringbuffers/Makefile]: TARGET/TARGET_BPF should be rebuilt for each selftest invocation.

**Comment by grantseltzer** [comment]
@itaysk You're right that the select statement causes dropped events. Back to the drawing board.

File: libbpfgo/selftest/ringbuffers/run.sh
Comment (grantseltzer) [reply libbpfgo/selftest/ringbuffers/run.sh:1]: Yes, I need to learn about bats, will this week

File: libbpfgo/selftest/ringbuffers/Makefile
Comment (itaysk) [reply libbpfgo/selftest/ringbuffers/Makefile]: why is that? is there a reason to rebuild something if it's source hasn't changed?

File: libbpfgo/selftest/ringbuffers/run.sh
Comment (itaysk) [reply libbpfgo/selftest/ringbuffers/run.sh:1]: it can be in a future PR

**Comment by grantseltzer** [comment]
Alright! I am cautiously optimistic that i've gotten it right this time. If we're calling Stop/Close on the ringbuffer there's no reason to block waiting for a consumer of events. I've replaced the stop signal send in `poll` with a  goroutine as to avoid blocking. This should avoid dropped events (at least in userspace) and blocking channel read/writes.

Take a look and tell me what you think. I'm going to be really happy once I put together benchmarks so we can put some metrics behind my thinking that this avoids blocks.

File: tracee-ebpf/tracee/tracee.go
Comment (grantseltzer) [reply tracee-ebpf/tracee/tracee.go:848]: Done :)

File: libbpfgo/selftest/ringbuffers/Makefile
Comment (grantseltzer) [reply libbpfgo/selftest/ringbuffers/Makefile]: I was mistaken, I thought it wasn't picking up changes in the bpf code, but that's what the TARGET_BPF is depending on. Removing the PHONY tags.

File: libbpfgo/selftest/ringbuffers/self.bpf.c
Comment (yanivagman) [inline libbpfgo/selftest/ringbuffers/self.bpf.c:38]: Indentation of this function (better use spaces as a convention like we do in tracee.bpf.c file)

**Review by yanivagman — APPROVED** [review]
LGTM!

File: libbpfgo/libbpfgo.go
Comment (itaysk) [inline libbpfgo/libbpfgo.go:689]: just a readability suggestion (totally optional) to move this stop condition to the for line
```
for !rb.stopped {
```

File: libbpfgo/libbpfgo.go
Comment (itaysk) [inline libbpfgo/libbpfgo.go:678]: just a readability suggestion (totally optional) to move this out of poll (into Start)

File: libbpfgo/libbpfgo.go
Comment (itaysk) [inline libbpfgo/libbpfgo.go:668]: not sure I understand: if Close was called before Stop, you wait for Stop to have happened? isn't it better to call Stop instead?

File: libbpfgo/libbpfgo.go
Comment (itaysk) [reply libbpfgo/libbpfgo.go:43]: ☝️

File: libbpfgo/libbpfgo.go
Comment (grantseltzer) [reply libbpfgo/libbpfgo.go:668]: Yes good point, need to add a check in Stop() to see if it's already stopped because of the nature of channels but this makes more sense

**Comment by grantseltzer** [comment]
@itaysk I realized why I wanted the make targets to be marked `.PHONY`, if there's changes to libbpfgo the makefile wasn't picking that up and therefore not recompiling.

File: libbpfgo/libbpfgo.go
Comment (grantseltzer) [reply libbpfgo/libbpfgo.go:668]: This throws off the fact that there's only one consumer of `rb.stop` so can cause a block. Now that I think of it, do we even need a `Stop` and a `Close`? Who is going to stop a ringbuffer and then restart it later? I'd be fine with just removing `Stop` and having the write to `rb.stop` in `Close()`. 

cc @yanivagman

Comment (itaysk) [reply libbpfgo/libbpfgo.go:668]: I'll let you and Yaniv decide if you want to support start after stop (make sure to document it) but certainly we don't want to support start after close, so I think either way close should just call stop, no?

**Comment by itaysk** [comment]
> @itaysk I realized why I wanted the make targets to be marked .PHONY, if there's changes to libbpfgo the makefile wasn't picking that up and therefore not recompiling.

You already have `$(TARGET): $(GO_SRC)` which should take care of this dependency, I don't see a reason to why it needs to be `.PHONY`, but I won't let it block the PR (we can fix the makefile later). If you want we could look into the makefile together lmk

**Comment by grantseltzer** [comment]
@itaysk maybe I misunderstand how .PHONY works, if I make a change in libbpfgo (two directories back), how `$(TARGET): $(GO_SRC)` affect rebuilding the Go binary?

**Comment by itaysk** [comment]
right (I thought GO_SRC included that), so I think the correct solution is to make TARGET depend on libbpfgo. I mean, regardless of the rebuild issue, it is correct to say that this test needs libbpfgo right? if we express the relationship correctly, make should be able to just do the right thing without workarounds like .PHONY. 
PHONY tells make that this target is not a real file, which is not a case here IMU

**Comment by grantseltzer** [comment]
@itaysk Ah, I understand. I've added a LIBBPFGO_SRC dependency to the BUILD target (not sure if my terminology is correct). I think this is what you were getting at.

File: libbpfgo/libbpfgo.go
Comment (itaysk) [reply libbpfgo/libbpfgo.go:43]: I still see whitespace issues in the commented C

File: libbpfgo/libbpfgo.go
Comment (itaysk) [inline libbpfgo/libbpfgo.go:660]: looks like the select is unnecessary?

File: libbpfgo/selftest/ringbuffers/Makefile
Comment (itaysk) [inline libbpfgo/selftest/ringbuffers/Makefile:29]: nit: probably should use TARGET and TRAGET_BPF here

**Review by itaysk — APPROVED** [review]
LGTM :)
there are some small comments here and in the previous review if you want to address/ignore before merging.
(and please squash before merging)

File: libbpfgo/libbpfgo.go
Comment (grantseltzer) [reply libbpfgo/libbpfgo.go:660]: This is what makes the channel write non-blocking.

File: libbpfgo/libbpfgo.go
Comment (grantseltzer) [reply libbpfgo/libbpfgo.go:43]: So strange, not sure why that kept happening...

**Pull request merged**
